node/src/components/consensus/highway_core/highway.rs

```diff
@@ -574,6 +571,20 @@ impl<C: Context> Highway<C> {
         evidence_effects.extend(self.on_new_unit(&unit_hash, now, rng));
         evidence_effects
     }
+
+    /// Adds endorsements to the state. If there are conflicting endorsements, `NewVertex` effects
+    /// are returned containing evidence to prove them faulty.
+    fn add_endorsements(&mut self, endorsements: Endorsements<C>) -> Vec<Effect<C>> {
+        let evidence = self
+            .state
+            .find_conflicting_endorsements(&endorsements, &self.instance_id);
+        self.state.add_endorsements(endorsements);
```

Comment:
Won't this add also the conflicting endorsements to the local state?

Reply:
Yes, and that's intended: An honest valiators might already created new units because of one of them.
Not adding (or even removing) certain endorsements sounds dangerous to me; we could discuss that in the consensus meetings, but for now I'd rather just add them all to the state.

Reply:
OK, maybe you're right. It sounds similar to detecting equivocation – we won't remove equivocating unit since an honest validator might have already created a unit that cites it.